charms/knative-serving/src/charm.py

```diff
@@ -70,6 +81,33 @@ def _on_install(self, _):
     def _on_config_changed(self, _):
         self._apply_and_set_status()
 
+    def _on_ingress_gateway_relation_changed(self, _) -> None:
+        """Sends the ingress gateway info through the gateway-info relation."""
+        try:
+            self._ingress_gateway_provider.send_gateway_relation_data(
+                gateway_name=self.model.config["istio.gateway.name"],
+                gateway_namespace=self.model.config["istio.gateway.namespace"],
+            )
+        except GatewayRelationMissingError:
+            # If there is no relation, we cannot send data. In reality, the execution
+            # should not get this far.
+            self.unit.status = BlockedStatus("Please add gateway-info relation.")
+            return
+
+    def _on_local_gateway_relation_changed(self, _) -> None:
+        """Sends the local gateway info through the gateway-info relation."""
+        try:
+            # FIXME: The local gateway name is hardcoded in the KnativeServing.yaml.j2
+            self._local_gateway_provider.send_gateway_relation_data(
+                gateway_name="knative-local-gateway",
+                gateway_namespace=self.model.config["namespace"],
+            )
+        except GatewayRelationMissingError:
+            # If there is no relation, we cannot send data. In reality, the execution
+            # should not get this far.
+            self.unit.status = BlockedStatus("Please add gateway-info relation.")
```

Comment:
will not unit go into blocked state on relation departed?

what if I want to change local to ingress ?

Reply:
> will not unit go into blocked state on relation departed?

Will RelationDeparted trigger a RelationChanged? I don't think it will.

> what if I want to change local to ingress ?

To clarify, you can use both or just one without issue. The requirer charm will relate to the same charm using different relations.

Reply:
> Will RelationDeparted trigger a RelationChanged? I don't think it will, but please correct me if I am wrong.

no, it will not

Reply:
so, from what you are saying we should not even reach this stage, ever

it is sort of extra security. Or?

Reply:
Yup, extra security, that is why I left the comment there